### PR TITLE
Interfaces by way of Unit Tests (Kinda)

### PR DESCRIPTION
**Pull request by TemporalOroboros**
## About The Pull Request
- Adds Interfaces (Sort of)

Interfaces are essentially datum types with vars and procs that other datum types may implement. If an interface has been declared with var "foo" and proc "bar". Any datum type that claims to implement the interface must also implement those vars and procs.

- [ ]TODO: Actually test this so that I know for certain it works as advertized.
- [ ]TODO: Add one or two examples of how to declare and implement interfaces.

### CODE REVIEW

**Comment by ninjanomnom** [comment]
This isn't really worth doing the checks at the dm level as it can't handle checking that a proc with an interface argument is actually being given arguments with that interface implemented.

Conceptually, getting in interfaces would be great but should be done via lint.

**Comment by TemporalOroboros** [comment]
I would love for them to be done via lint. The difficulty is checking them at runtime, for filtering lists and such.

**Comment by ninjanomnom** [comment]
Sure, we'd want some dm side code to handle interfaces as well, but using interfaces with *just* the dm side code isn't really a great idea.

File: code/__DEFINES/interfaces.dm [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#define IMPLEMENTS_INTERFACE(typepath, interfaces...) \
+/datum/controller/subsystem/interfaces/aggregate_implementations() { \
+	. = ..(); \
+	for(var/interface in list(##interfaces)) { \
+		.[interface][##typepath] = TRUE; \
+	} \
+}
+
+/// Converts something that is not a string to a string.
+#define NOT_A_STRING_TO_STRING(not_a_string) (#not_a_string)
```
Comment (Mothblocks) [inline code/__DEFINES/interfaces.dm:12]: Name it `STRINGIFY`, a more common and obvious name.

File: code/__DEFINES/interfaces.dm [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
...takearmsagainstaseaoftroublesandbyopposingendthemtodietosleepnomoreandbyasleeptosayweendtheheartacheandthethousandnaturalshocksthatfleshisheirtotisaconsummationdevoutlytobewishdtodietosleeptosleepperchancetodreamaytherestherubforinthatsleepofdeathwhatdreamsmaycomewhenwehaveshuffledoffthismortalcoilmustgiveuspausetherestherespectthatmakescalamityofsolonglifeforwhowouldbearthewhipsandscornsoftimethoppressorswrongtheproudmanscontumelythepangsofdisprizdlovethelawswowyouareactuallyreadingthisgoodforyoudelaytheinsolenceofofficeandthespurnsthatpatientmeritofthunworthytakeswhenhehimselfmighthisquietusmakewithabarebodkinwhowouldfardelsbeartogruntandsweatunderawearylifebutthatthedreadofsomethingafterdeaththeundiscoveredcountryfromwhosebournnotravellerreturnspuzzlesthewillandmakesusratherbearthoseillswehavethanflytoothersthatweknownotofthusconsciencedothmakecowardsofusallandthusthenativehueofresolutionissickliedoerwiththepalecastofthoughtandenterprisesofgreatpithandmomentwiththisregardtheircurrentsturnawryandlosethenameofaction________________________________________________________________________________________________________________________________________________________________________
```
Comment (Mothblocks) [inline code/__DEFINES/interfaces.dm:14]: No.

File: code/datums/interfaces/_interfaces.dm [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+/datum/interface
+	/// The var used to hold all the procs declared by this interfaces. Because an equivalent to .vars does not exist for procs.
+	var/list/INTERFACE_PROC_CACHE_NAME
+
+/datum/interface/New()
+	. = ..()
+	INTERFACE_PROC_CACHE_NAME = ____aggregate_procs()
+
+/// Aggregates a list of all of the procs this interface declares.
+/datum/interface/proc/____aggregate_procs()
```
Comment (Mothblocks) [inline code/datums/interfaces/_interfaces.dm:18]: You give it 4 underscores, seemingly to say "I'm SUPER SUPER SUPER SUPER sure you don't want this", then auto-doc it anyway? 

I really don't like this naming style. I say do `_aggregate_procs` at *best*.

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Agreed with ninjanomnom, but here's some quick tired review.

**Comment by Mothblocks** [comment]
If you want a good example of something that would benefit from explicit interfaces, it's `/datum/client_interface`. Currently, to my knowledge, it only has `var/datum/prefs/prefs`, but in theory it could have more later on, which increases the chance it will drift.

**Comment by ninjanomnom** [comment]
I've thought about this since and I can't see this being worth enough on its own. If interfaces are ever done with the linter or otherwise then this will be wanted, but as it is this isn't something we can support.

**Pull request closed without merging**